Telnet-style reporter and per-metric tags

alugowski opened the pull request:
I'm asending our internal changes to your reporter upstream. Our pipeline requires the older Telnet-style reporter, and we felt others might as well. Also, TSDB supports per-metric tags, something that seems to be missing from every CodaHale metrics reporter I've found. While there's no clean solution to this problem due to the limitations of CH Metrics' naming scheme, the approach we introduced in this PR works well for us and it's purely a superset of the existing design.

* Telnet-style reporter option that can write to a Socket or a Writer
* Per-metric tag support
* Enable or disable appending of '.count' and '.value' to Counters and Gauges
* Tests for the above
* Fixed JavaDoc warnings

sps commented:
Hey @alugowski - this is generally good with me. Can you please resolve the findbugs issues that is causing your build to fail?

```
[INFO] Format string should use %n rather than \n in com.github.sps.metrics.opentsdb.OpenTsdbMetric.toTelnetPutString() ["com.github.sps.metrics.opentsdb.OpenTsdbMetric"] At OpenTsdbMetric.java:[lines 32-328]
[INFO] Found reliance on default encoding in com.github.sps.metrics.opentsdb.OpenTsdbTelnet$SocketWriterFactory.getWriter(): new java.io.OutputStreamWriter(OutputStream) ["com.github.sps.metrics.opentsdb.OpenTsdbTelnet$SocketWriterFactory"] At OpenTsdbTelnet.java:[lines 59-71]
```

alugowski commented:
Hey @sps I fixed those two findbugs instances. I still don't know why an INFO message would block a merge. I fixed it to make Travis happy, but I'm only assuming that this will work with OpenTSDB alright. I personally would stay with '\n' on all platforms.

I also don't know why it's complaining about a merge conflict. When I look at the diff it does not show any conflicts.

sps commented:
Hey @alugowski thanks for resolving those issues. 
It's a fair point about '\n' in this particular use case.

Here is the merge conflict

```diff
@@@ -91,7 -88,14 +91,18 @@@ public class OpenTsdb
          }
      }

++<<<<<<< HEAD
 +    private OpenTsdb(WebTarget apiResource) {
++=======
+       /**
+        * For OpenTsdbTelnet
+        */
+       protected OpenTsdb() {
+               this.apiResource = null;
+       }
+
+     private OpenTsdb(WebResource apiResource) {
++>>>>>>> 33eca7ed7aca435b649444e4b94e679349613053
          this.apiResource = apiResource;
```

sps commented:
Hi @alugowski - any plans to resolve the merge conflicts?

csueiras commented:
I took this work and resolved the conflicts in PR: https://github.com/sps/metrics-opentsdb/pull/17
